Fix Asynchronous Dispatch Logic in AwsAsyncContext with Spring's DispatcherServlet

2012160085 opened the pull request:
*Issue #, if available:  #630 

*Description of changes: 

## Problem

While examining #630, I found an issue within the asynchronous dispatch logic of `AwsAsyncContext` when interacting with Spring's `DispatcherServlet`.

* **Premature Re-dispatch**: Currently, `AwsAsyncContext` prematurly triggers re-dispatch. 
When handling asynchronous requests with `DispatcherServlet`,  `DispatcherServlet`'s`doDispatch` should be called to complete the initial request. Only after the initial request concludes, should the `doDispatch` be called again for re-dispatch. However, in the present implementation, this sequence is disrupted. During the processing of the initial request by the `doDispatch` method, the `dispatch` method of `AwsAsyncContext` is invoked, which in turn directly calls the `doFilter` method for immediate re-dispatch. This leads to premature execution of the re-dispatch before the post-processing logic of `doDispatch`(which is handling the initial request) is executed.

* **EntityManager Bind Timing Issue**: While using Spring JPA, due to the aforementioned issue, an error arises when trying to bind the `EntityManager` to threadLocal during the preHandle phase of the re-dispatch. This is because the `EntityManager` is already bound from the initial request, and the post-processing logic meant to unbind it hasn't been executed yet. As a result, attempting to bind it again triggers the following error:

```bash
Value [org.springframework.orm.jpa.EntityManagerHolder@38a6ca39] already bound to key [org.springframework.orm.jpa.LocalContainerEntityManagerFactoryBean@39296cef].
```

## Changes

To resolve this, I made modifications to the `SpringBootLambdaContainerHandler::handleRequest` and `AwsAsyncContext::dispatch` by referencing the way requests are handled in `apache.coyote.AbstractProcessorLight::process`.
The snippet for the proposed solution is as follows:

```java
protected void handleRequest(HttpServletRequest containerRequest, AwsHttpServletResponse containerResponse, Context lambdaContext) throws Exception {
    // ... [other code]
    doFilter(containerRequest, containerResponse, reqServlet);
    //added addtional filter logic
    if(requiresAsyncReDispatch(containerRequest)) {
        doFilter(containerRequest, containerResponse, reqServlet);
    }
    // ... [other code]
}
```

```java
public void dispatch() {
    // ... [other code]
    //added dispatchStarted flag
    if (!dispatchStarted.get()) {
        dispatchStarted.set(true);
    } else {
        dispatched.set(true);
    // removed the line that directly calls doFilter
        notifyListeners(NotificationType.START_ASYNC, null);
    }
    // ... [other code]
}
```

By ensuring that the post-processing of the initial request is complete and only then re-invoking the doFilter when necessary, we can bypass the issues mentioned above.

I have not yet added test cases, and there are two test cases marked with @Disable since AwsAsyncContext no longer maps requests to the servlet. I think we can talk about this solution or potentially explore better approaches.


By submitting this pull request

- [x] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [x] I confirm that I've made a best effort attempt to update all relevant documentation.

mbfreder commented:
Thanks for your contribution. I will review this today and let you know.

mbfreder commented:
Looks good to me. Could you add tests, please?

2012160085 commented:
I added test cases

deki commented:
How do we plan to deal with the `@Disabled` test cases? I've not worked much with WebFlux but we need to make sure that the things implemented in #239 are still working.

2012160085 commented:
> How do we plan to deal with the `@Disabled` test cases? I've not worked much with WebFlux but we need to make sure that the things implemented in #239 are still working.

Thank you for taking the time to review.

The two `@Disabled` test cases are testing whether `AwsAsyncContext`'s `dispatch` correctly maps to the right servlet.
Specifically, it's testing [this line](https://github.com/awslabs/aws-serverless-java-container/blob/0fedd837516a0329b15585e012fd715554b2794a/aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsAsyncContext.java#L77C13-L77C13) to see whether the right servlet is input when calling `doFilter`.

However, in my commit, this line has removed from the `dispatch` function, and the responsibility for mapping the servlet has now shifted from `AwsAsyncContext` to `SpringBootLambdaContainerHandler`.

I believe the disabled cases are adequately covered in the test cases of `aws-serverless-java-container-springboot` and `springboot3`. If necessary, we can add test cases about servlet mapping for asynchronous requests.

However, as you pointed out, I seem to have overlooked testing whether the request path is properly amended when `dispatch` is called in `AwsAsyncContextTest`. I've added that test.

2012160085 commented:
I missed that both `springboot3` and `spring` handle async requests the same way. After looking over it again, I've updated the `SpringLambdaContainerHandler` in `aws-serverless-java-container-spring` and added some tests.

2012160085 commented:
@deki Any further feedback on my update?

deki commented on aws-serverless-java-container-core/src/test/java/com/amazonaws/serverless/proxy/internal/servlet/AwsAsyncContextTest.java:
```diff
@@ -58,6 +73,7 @@ void dispatch_sendsToCorrectServlet() {
         assertEquals(202, handler.getResponse().getStatus());
     }
 
+    @Disabled("AwsAsyncContext does not sends to servlet anymore")
```
In my view we should remove the two disabled test cases as nobody will re-enable them in the future.

2012160085 replied:
Removed `AwsLambdaServletContainerHandler` from `AwsAsyncContext` and these two Disabled tests.

deki reviewed (changes requested):
`AwsAsyncContext` no longer depends on the `AwsLambdaServletContainerHandler` could you please remove the unused variable as well?

deki commented:
Hey @2012160085, 
sorry for the delay, I saw your comment on #630 and had therefore de-prioritized this PR. 

I read https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#asynchronous-processing, looked at https://github.com/apache/tomcat/blob/main/java/org/apache/catalina/core/AsyncContextImpl.java and noticed we may required additional fixes. The async processing capability should work within aws-serverless-java-container-core but with the proposed change you made calling filters exclusive to -spring and -springboot3. Filters for apps running -jersey and other potential implementations would no longer be called or am I missing something?

2012160085 commented:
> Hey @2012160085, sorry for the delay, I saw your comment on #630 and had therefore de-prioritized this PR.
> 
> I read https://jakarta.ee/specifications/servlet/6.0/jakarta-servlet-spec-6.0#asynchronous-processing, looked at https://github.com/apache/tomcat/blob/main/java/org/apache/catalina/core/AsyncContextImpl.java and noticed we may required additional fixes. The async processing capability should work within aws-serverless-java-container-core but with the proposed change you made calling filters exclusive to -spring and -springboot3. Filters for apps running -jersey and other potential implementations would no longer be called or am I missing something?

You seem to be correct based on what I've reviewed. I will take a closer look into this

deki commented:
@2012160085 just a heads up, we are planning to release 2.0.0 at the end of this month. are you still working on it so we can include it?

2012160085 commented:
> @2012160085 just a heads up, we are planning to release 2.0.0 at the end of this month. are you still working on it so we can include it?

I haven't been able to focus on this recently, but I will make sure to wrap it up by the end of the month.

2012160085 commented:
> @2012160085 just a heads up, we are planning to release 2.0.0 at the end of this month. are you still working on it so we can include it?

I think I've wrapped it up.
* Modified the serverless-java-container-core to revise the ReDispatch logic. 
I relocated the ReDispatch logic from springboot3 and spring to the serverless-java-container-core, ensuring that the ReDispatch logic now operates exclusively within the serverless-java-container-core.
* Added a failing case, jpaAppTest, to serverless-java-container-springboot3. Additionally, I added asynchronous request test cases to Spring.

valerena commented on aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsAsyncContext.java:
```diff
@@ -68,16 +68,15 @@ public boolean hasOriginalRequestAndResponse() {
+        if (!dispatchStarted.get()) {
+            dispatchStarted.set(true);
+        } else {
             dispatched.set(true);
-            handler.doFilter(req, res, ((AwsServletContext)req.getServletContext()).getServletForPath(req.getRequestURI()));
             notifyListeners(NotificationType.START_ASYNC, null);
-        } catch (ServletException | IOException e) {
-            notifyListeners(NotificationType.ERROR, e);
         }
     }
```
We could improve this a little bit, using functionality specific to AtomicBoolean, to eventually prevent a race condition (although I'm not sure how important it is here). 

But this is equivalent to:

```
       if (dispatchStarted.getAndSet(true)) {
           dispatched.set(true);
           notifyListeners(NotificationType.START_ASYNC, null);
       }
```

You probably looked at this code more times than me, but let me know if this makes sense to you.

2012160085 replied:
It seems better to change the code as you suggested, for the sake of atomicity.

deki commented:
Thanks again for your contribution, @2012160085!
